### PR TITLE
Missions that don't block minors

### PR DESCRIPTION
**Pull request by warp-core**
**Feature**

## Summary
Currently, a mission marked "minor" can only offer one at a time and when no other mission is being offered.
That is, if any non-"minor" missions are being offered, no "minor" missions will be offered, and if multiple "minor" missions may be offered, and no non-"minor" missions, all but the last "minor" mission will be discarded, and only that mission will be offered.

This can be somewhat limiting, so this PR introduces a way to mark a mission as "non-blocking". Such missions will not prevent "minor" missions from being offered; "minor" missions can offer alongside them.

The most obvious use for this would be the spaceport reminder resetter.
The spaceport reminder missions currently work as follows:
1. The "Spaceport Reminder Setter" mission offers on landing (in practice, this will be when the pilot is first created) and sets conditions storing the current month and year values ("spaceport reminder month" and "spaceport reminder year").
2. The "AA Spaceport Reminder Resetter" mission is a "minor" spaceport mission. Because its name begins with "AA", it will not offer unless no other vanilla missions are going to offer; every vanilla "minor" mission will take precedence over this one and all non-"minor" missions will block it (because it is "minor"). (A plugin could define a new "minor" mission that would have a lower precedence than this one, if it really wanted.) When this mission is offered, it updates the conditions set in the previous mission with the current month and year. This mission will stop offering once the "Spaceport Reminder" mission has been offered or the player has "chosen sides" in the FW story.
3. The "Spaceport Reminder" mission will offer on landing approximately seven months after the last time the "Resetter" mission ran, or after the "Setter" mission, if the "Resetter" has never run. It does not offer once the player has "chosen sides" in the FW story.

The purpose of these missions is, as the name suggests, to remind the player to visit the spaceport (to pick up main story missions). The idea being that if you do not visit the spaceport very often or at all, or have not visited it recently, the "Resetter" mission will not have run recently, so the "Spaceport Reminder" mission will be offered, providing the reminder conversation.

However, the current implementation requires that you visit the spaceport and not be offered any other missions for the "Resetter" to run. This means that it will not be reset when you are offered a mission. It is perhaps unlikely that a player would be offered a mission on every visit to the spaceport in seven months of in game time, assuming they visit the spaceport every time they land, but even somewhat less frequent visitations probably don't merit a reminder.

The feature in this PR could be used to allow the "Resetter" mission to run every time the player visits the spaceport regardless of if the player is offered another mission and without blocking "minor" missions.

## Screenshots
N/A

## Usage examples
```
mission
    "non-blocking"
```

## Testing Done
Literally nothing. It might not even compile.

## Save File
This save file can be used to test these changes:
Maybe another time.

## Artwork Checklist
N/A

## Wiki Update
soon:tm:

## Performance Impact
Immeasurable, probably.

### CODE REVIEW

**Comment by TheGiraffe3** [comment]
Would every vanilla mission have to be marked "non-minor" in this case?

**Comment by warp-core** [comment]
> Would every vanilla mission have to be marked "non-minor" in this case?

In which case?

**Comment by TheGiraffe3** [comment]
> The feature in this PR could be used to allow the "Resetter" mission to run every time the player visits the spaceport regardless of if the player is offered another mission and without blocking "minor" missions.

_could be used_: to be used, would we have to give every vanilla mission the new tag?

**Comment by Amazinite** [comment]
No. "non-blocking" means the mission itself doesn't block minor missions from offering. Since the behavior we want is that the resetter mission always offers and does not block other missions, we only need to give the resetter mission "non-blocking" while removing the "minor" tag. No other missions need affected.

File: source/PlayerInfo.cpp [hunk 1]
```diff
@@ -4009,6 +4009,7 @@ void PlayerInfo::CreateMissions()
 	// Check for available missions.
 	bool skipJobs = planet && !planet->GetPort().HasService(Port::ServicesType::JobBoard);
 	bool hasPriorityMissions = false;
+	unsigned nonBlockingMissions = 0;
```
Comment (petervdmeer) [inline source/PlayerInfo.cpp:4012]: I'd expect non-blocking missions to just always be offered (regardless of anything else), and not blocking anything else. So don't let them run in parallel with minor missions, but just let them run in parallel with everything (or in this code, run them ahead of any other mission).

**Comment by petervdmeer** [comment]
> No. "non-blocking" means the mission itself doesn't block minor missions from offering. Since the behavior we want is that the resetter mission always offers and does not block other missions, we only need to give the resetter mission "non-blocking" while removing the "minor" tag.

I agree with this reasoning. But that also means that a mission is either major, or minor or non-blocking. Missions should not have multiple of those keywords (they seem to be mutually exclusive).

File: source/PlayerInfo.cpp (hunk 1 above)
Comment (petervdmeer) [reply source/PlayerInfo.cpp:4012]: As discussed on discord; this code is indeed correct and needed to ensure that minor missions still offer.

**Review by petervdmeer — APPROVED** [review]

**Comment by petervdmeer** [comment]
> I agree with this reasoning. But that also means that a mission is either major, or minor or non-blocking. Missions should not have multiple of those keywords (they seem to be mutually exclusive).

We can handle the mutually exclusive parts (and the soon^tm Wiki update) in a follow-up PR. This PR by itself is already an improvement.

**Pull request merged**